pkg/catalog/manager/manager_test.go

```diff
@@ -0,0 +1,81 @@
+package manager
+
+import (
+	"testing"
+
+	"github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLatestAvailableTemplateVersion(t *testing.T) {
+	template := &v3.CatalogTemplate{
+		Template: v3.Template{
+			Spec: v3.TemplateSpec{
+				Versions: []v3.TemplateVersionSpec{
+					{
+						ExternalID:        "catalog://?catalog=library&template=artifactory-ha&version=0.12.16",
+						Version:           "0.12.16",
+						RancherMinVersion: "v2.2.0",
+						RancherMaxVersion: "v2.3.0",
+					},
+					{
+						ExternalID:        "catalog://?catalog=library&template=artifactory-ha&version=0.12.15",
+						Version:           "0.12.15",
+						RancherMinVersion: "v2.1.0",
+						RancherMaxVersion: "v2.2.0",
+					},
+					{
+						ExternalID:        "catalog://?catalog=library&template=artifactory-ha&version=0.12.14",
+						Version:           "0.12.14",
+						RancherMinVersion: "v2.0.0",
+						RancherMaxVersion: "v2.1.0",
+					},
+				},
+			},
+		},
+	}
+
+	templateWithoutRancherVersion := &v3.CatalogTemplate{
+		Template: v3.Template{
+			Spec: v3.TemplateSpec{
+				Versions: []v3.TemplateVersionSpec{
+					{
+						ExternalID: "catalog://?catalog=library&template=artifactory-ha&version=0.12.16",
+						Version:    "0.12.16",
+					},
+					{
+						ExternalID: "catalog://?catalog=library&template=artifactory-ha&version=0.12.15",
+						Version:    "0.12.15",
+					},
+					{
+						ExternalID: "catalog://?catalog=library&template=artifactory-ha&version=0.12.14",
+						Version:    "0.12.14",
+					},
+				},
+			},
+		},
+	}
+
+	testLatestAvailableTemplateVersion(t, "v2.1.0", "0.12.15", template)
```

Comment:
Build up a map of your test cases then in your test func loop over the map and process the cases
In the test func you can create one assert at the top with assert.New(t)

Reply:
Oh this already existed...